pilot/pkg/bootstrap/config_compare.go

```diff
@@ -15,22 +15,39 @@
 package bootstrap
 
 import (
+	"reflect"
 	"strings"
 
 	gogoproto "github.com/gogo/protobuf/proto" // nolint: depguard
 	"google.golang.org/protobuf/proto"
+	gateway "sigs.k8s.io/gateway-api/apis/v1"
 
 	"istio.io/istio/pkg/config"
 )
 
 // needsPush checks whether the passed in config has same spec and hence push needs
 // to be triggered. This is to avoid unnecessary pushes only when labels have changed
 // for example.
-func needsPush(prev config.Config, curr config.Config) bool {
+func needsPush(prev config.Config, curr config.Config) (toPush bool, skipGenXDS bool) {
 	if prev.GroupVersionKind != curr.GroupVersionKind {
 		// This should never happen.
-		return true
+		return true, false
+	}
+
+	if prev.GroupVersionKind.Group == gateway.GroupName {
+		// We always push for Gateway API resources so that the status can be repaired.
+		return true, needsPushXDSForGatewayAPI(&prev, &curr)
 	}
+
+	return needsPushForOther(&prev, &curr), false
+}
+
+func needsPushXDSForGatewayAPI(prev *config.Config, curr *config.Config) bool {
+	// We skip the xDS push for Gateway API when the Spec is the same.
+	return reflect.DeepEqual(prev.Spec, curr.Spec)
```

Comment:
This should be negated right? needsPush should return true if the spec is not equal

Reply:
I am not sure I am comfortable adding this DeepEqual on the hotpath like this.

Reply:
I do a benchmark and we can handle 4 million DeepEqual in one second:
```
package main

import (
	"reflect"
	"testing"

	istioapi "istio.io/api/networking/v1alpha3"
	istiov1a3 "istio.io/client-go/pkg/apis/networking/v1alpha3"
)

func deepEq(a *istiov1a3.Gateway, b *istiov1a3.Gateway) bool {
	return reflect.DeepEqual(a.Spec, b.Spec)
}

func BenchmarkBase(b *testing.B) {
	gw := &istiov1a3.Gateway{
		Spec: istioapi.Gateway{
			Servers: []*istioapi.Server{
				{
					Hosts: []string{"*.example.com"},
					Port: &istioapi.Port{
						Number:   80,
						Name:     "http",
						Protocol: "HTTP",
					},
				},
				{
					Hosts: []string{"*.example.net"},
					Port: &istioapi.Port{
						Number:   80,
						Name:     "http",
						Protocol: "HTTP",
					},
				},
				{
					Hosts: []string{"*.example.org"},
					Port: &istioapi.Port{
						Number:   80,
						Name:     "http",
						Protocol: "HTTP",
					},
				},
				{
					Hosts: []string{"*.example"},
					Port: &istioapi.Port{
						Number:   80,
						Name:     "http",
						Protocol: "HTTP",
					},
				},
			},
			Selector: map[string]string{
				"istio": "ingressgateway",
			},
		},
	}
	for i := 0; i < b.N; i++ {
		deepEq(gw, gw)
	}
}

```

```
go test -v -cpu=1 -run=none -bench=. -benchtime=1s                        
goos: darwin
goarch: arm64
pkg: example.com/go
BenchmarkBase
BenchmarkBase 	4433490	      265.9 ns/op
PASS
ok  	example.com/go	2.663s
```

Each Gateway is 555 bytes represented in yaml, so we can handle 2GB configuration in a second.